pyteal/ast/scratchvar.py

```diff
@@ -47,5 +54,86 @@ def load(self) -> ScratchLoad:
         """Load value from Scratch Space"""
         return self.slot.load(self.type)
 
+    def index(self) -> Expr:
+        return self.slot.index()
+
 
 ScratchVar.__module__ = "pyteal"
+
+
+class DynamicScratchVar(ScratchVar):
+    """
+    Example of Dynamic Scratch space whereby the slot index is picked up from the stack:
+        .. code-block:: python1
+
+            player_score = DynamicScratchVar(TealType.uint64)
+
+            wilt = ScratchVar(TealType.uint64, 129)
+            kobe = ScratchVar(TealType.uint64)
+            dt = ScratchVar(TealType.uint64, 131)
+
+            seq = Seq(
+                player_score.set_index(wilt),
+                player_score.store(Int(100)),
+                player_score.set_index(kobe),
+                player_score.store(Int(81)),
+                player_score.set_index(dt),
+                player_score.store(Int(73)),
+                Assert(player_score.load() == Int(73)),
+                Assert(player_score.index() == Int(131)),
+                player_score.set_index(wilt),
+                Assert(player_score.load() == Int(100)),
+                Assert(player_score.index() == Int(129)),
+                Int(100),
+            )
+    """
+
+    def __init__(self, ttype: TealType = TealType.anytype):
+        """Create a new DynamicScratchVar which references other ScratchVar's
+
+        Args:
+            ttype (optional): The type that this variable can hold. Defaults to TealType.anytype.
+        """
+        self.type = ttype
+        self._indexer = ScratchVar(TealType.uint64)
```

Comment:
_Assuming_ `DynamicScratchVar` continues extending `ScratchVar`, the proposal makes sense to me.

* I had planned to ask if it's intentional that `super().__init__` is _not_ called.  The proposal eliminates the ambiguity.  And I think the proposal provides equivalent functionality.
* I say _assuming_ because I'm still evaluating the class hierarchy.  I think I understand how https://github.com/algorand/pyteal/pull/198#discussion_r810271366 motivates inheritance.  But, it's unclear to me if that's the only viable path.  
  * Don't put much weight on my skepticism.  It _probably_ is the appropriate choice.  I have gaps in my intuition.

Reply:
I intentionally did not call `__super()__` as I wanted to maintain the compositional aspect of `DynamicScratchVar`. Am not sure yet what the implications are of Jason's proposal is. This is another good talking point.

Reply:
👍